Switch to MongoDB

autumnfound opened the pull request:
Added MongoDB support. Added random select, a few filters, BSON codecs,
sorting, secret properties.

Change-Id: I613403d5586f192fd68666df234139353069f8a2
Signed-off-by: Martin Lowe <martin.lowe@eclipse-foundation.org>

autumnfound commented:
There are still superficial changes to make in these branches (comments, tidying up some code), but I wanted to create this PR to get eyes on general structure before I commit too far to the current layout. Some of the things I think are the bigger question marks are the SortableField annotations and the combintation of codecs and converters in the DTO package

mbarbero commented on src/main/java/org/eclipsefoundation/marketplace/config/SecretConfigSource.java:
```diff
@@ -0,0 +1,99 @@
+			}
+
+			// read each of the lines of secret config that should be added
+			try (FileReader reader = new FileReader(f); BufferedReader br = new BufferedReader(reader)) {
```
I guess that using the java.util.Properties class would avoid custom parsing code.

autumnfound replied:
I attempted to use Properties earlier, but whenever I would return the internal map, it wouldn't properly have read the file, and would cause issues. I can attempt to work it a little more if I can find the root cause of the issue

autumnfound replied:
It just seemed to work this time. Might be that I reused the buffered reader and it worked better than the basic file input stream I used last time.

mbarbero commented on src/main/java/org/eclipsefoundation/marketplace/helper/StreamHelper.java:
```diff
@@ -0,0 +1,49 @@
+	 * @return the result, or null if one couldn't be found
+	 */
+	public static <T> T awaitCompletionStage(CompletionStage<T> stage) {
+		CompletableFuture<T> future = stage.toCompletableFuture();
```
a CompletionStage is a CompletableFuture implementation, so no need to do that here.

autumnfound replied:
CompletionStage is the interface, and Quarkus also has implementations of the class that aren't straight casts which could cause issues.

autumnfound commented:
I've actioned your comments, and added a few more things, mainly the following:
- Added maintenance flag to MongoDB
- Added a MaintenanceException, and a Mapper that catches it in the resource.
- Created a few basic health (liveness/readiness) checks for the application. There is a new interface with a default implementation and 1 method to implement to standardize checking bean health.
I'll stop touching this now, as its in a stable state for alpha release.

mbarbero commented on src/main/java/org/eclipsefoundation/marketplace/config/SecretConfigSource.java:
```diff
@@ -0,0 +1,84 @@
+		if (secrets == null) {
+			this.secrets = new HashMap<>();
+			// use the class loader to get the secret.properties file path
+			URL fURL = getClass().getClassLoader().getResource("secret.properties");
```
Getting the secret from the classloader means that at one point you will package the secret along with binary. This is not a good security practice. The path to the secret properties file should also be a config. Use @Inject and document that this config should be pass as a system property at startup (-DsecretConfigPath=/path/to/secret)

autumnfound replied:
Because of the nature of the config using service loader, I don't have bean context so I can't use injection. I'll have to make do with `System.getProperty(name)` if I want to set that variable at runtime

autumnfound replied:
You're very right, it's begging to be a vector for leaking secret information. I've switched over locally to use the above call with a property name of `config.secret.path` to match the naming convention of other properties in the system.

mbarbero commented on src/main/java/org/eclipsefoundation/marketplace/config/SecretConfigSource.java:
```diff
@@ -0,0 +1,84 @@
+			}
+
+			// read each of the lines of secret config that should be added
+			try (FileReader reader = new FileReader(f); BufferedReader br = new BufferedReader(reader)) {
```
Do not split declaration of cascading "closeable". Please use instead:

try (BufferedReader br = new BufferedReader(new FileReader(f)))

It's the responsibility of BufferedReader.close to call FileReader.close properly.

mbarbero commented on src/main/java/org/eclipsefoundation/marketplace/config/SecretConfigSource.java:
```diff
@@ -0,0 +1,84 @@
+			this.secrets = new HashMap<>();
+			// use the class loader to get the secret.properties file path
+			URL fURL = getClass().getClassLoader().getResource("secret.properties");
+			if (fURL == null) {
```
You may want to log an info or a warning that secret url cannot be found.

mbarbero commented on src/main/java/org/eclipsefoundation/marketplace/config/SecretConfigSource.java:
```diff
@@ -0,0 +1,84 @@
+	@Override
+	public String getValue(String propertyName) {
+		// if secrets not found, retrieve them
+		if (secrets == null)
```
the == null test is already handled in getProperties(). Don't be smart here and just do

return this.getProperties().get(propertyName);

It makes the method much easier to read and not dependent on a method to have invisible side effect (ie. set this.secrets) on this object.

autumnfound commented:
@mbarbero I've pushed a large update that does a number of things within the project. I've documented them in the commit (which has split from the commit I was working on for base Mongo functionality). 

The reason I point this out is I've hit a blocker in that CDI is fighting me tooth and nail to create something in the RequestScope of Quarkus ([3. Scopes and contexts -> @RequestScoped](https://quarkus.io/guides/cdi-reference#supported_features)). The class I'm trying to manage through CDI is `MongoQuery`, which I can create manually in my ResourceListing class (see commented out code in said class for how I was doing it), but not when I set it to be managed. 

The items all individually work, its when I put it all together that it falls apart and complains that there are unresolved dependencies. I've included a stack trace that I get when attempting to build. The only thing I can think of is that type erasure interferes with the retrieval of the resources in CDI, but I don't know how to either verify or circumvent the issue. Do you have any thoughts?

```14:46:20,725 ERROR [io.qua.dev.DevModeMain] Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.eclipsefoundation.marketplace.model.MongoQuery<org.eclipsefoundation.marketplace.dto.Listing> and qualifiers [@Default]
	- java member: org.eclipsefoundation.marketplace.resource.ListingResource#mq
	- declared on CLASS bean [types=[org.eclipsefoundation.marketplace.resource.ListingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.eclipsefoundation.marketplace.resource.ListingResource]
	at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:835)
	at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:214)
	at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:106)
	at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:249)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:768)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:415)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1535)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1426)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

mbarbero commented:
The issue seems to be in the ARC component of Quarkus which AFAIK is the partial CDI implementation from Quarkus (made to better work with SubstrateVM) so there may be limitations. There is nothing obvious to me from a quick look. I'd suggest you open a question on stackoverflow as specified here https://quarkus.io/community/#discussions.
